flang/lib/Lower/OpenMP.cpp

```diff
@@ -435,7 +443,8 @@ static void genOMP(Fortran::lower::AbstractConverter &converter,
       if (std::get_if<Fortran::parser::OmpClause::Nowait>(&clause.u))
         wsLoopOp.nowaitAttr(firOpBuilder.getUnitAttr());
   }
-  createBodyOfOp<omp::WsLoopOp>(wsLoopOp, firOpBuilder, currentLocation);
+
+  createBodyOfOp<omp::WsLoopOp>(wsLoopOp, converter, currentLocation, iv);
```

Comment:
The worksharing loop iteration variable's value is undefined outside the loop (similar to unordered loop), it is OK to not store back the value. If it is marked lastprivate then it might need to be stored back.
Modifying the induction variable inside the loop is incorrect. It is OK to observe the value by passing a reference, but each thread gets its own copy of the variable.

Reply:
Yes, `fir.do_loop` has inclusive bounds and its iterations can count up or down. I don't know about WsLoop, but we'll need to be careful if the semantics are different. (Also, `fir.do_loop` is lowered in a way that the iteration count cannot be changed within the loop, even (hopefully) if someone tries backdooring.)

Reply:
omp.wsloop can be inclusive or exclusive based on an attribute. It can count up or down, but the lowering is currently there only for  up-counting loops. Modifying the iteration is not allowed in a worksharing loop.